source/MaterialXGenShader/ShaderStage.h

```diff
@@ -0,0 +1,303 @@
+#ifndef MATERIALX_SHADERSTAGE_H
+#define MATERIALX_SHADERSTAGE_H
+
+#include <MaterialXCore/Library.h>
+#include <MaterialXCore/Node.h>
+#include <MaterialXGenShader/Syntax.h>
+#include <MaterialXGenShader/ShaderGraph.h>
+#include <MaterialXGenShader/GenOptions.h>
+
+#include <queue>
+#include <sstream>
+#include <unordered_map>
+
+/// Macro for being/end of statements to be picked up by a given shader stage.
+/// For shaders that are multi-stage all code generation statements adding code 
+/// to the shader should be wrapped inside such begin/end stating its target.
+#define BEGIN_SHADER_STAGE(stage, name) if (stage.getName() == name) {
+#define END_SHADER_STAGE(stage, name) }
+
+/// Private macro to name the main shader stage.
+/// This is only used to handle static initialization order issues.
+/// API users should use the string identifyer MaterialX::MAIN_STAGE
+/// instead of this macro.
+#define _MAIN_STAGE_NAME "main"
+
+namespace MaterialX
+{
+
+/// Identifier for main shader stage.
+extern const string MAIN_STAGE;
```

Comment:
Can this just be a class static? MaterialX::MAIN_STAGE doesn't really have an global meaning.